src/meta/src/barrier/progress.rs

```diff
@@ -52,20 +78,27 @@ impl Progress {
         Self {
             states,
             done_count: 0,
+            progress: 0.0,
+            creating_mv_id,
+            upstream_mv_count,
+            upstream_total_key_count,
+            definition,
         }
     }
 
     /// Update the progress of `actor`.
-    fn update(&mut self, actor: ActorId, new_state: ChainState) {
+    fn update(&mut self, actor: ActorId, new_state: ChainState, upstream_total_key_count: u64) {
+        self.upstream_total_key_count = upstream_total_key_count;
         match self.states.get_mut(&actor).unwrap() {
-            state @ (ChainState::ConsumingSnapshot | ChainState::ConsumingUpstream(_)) => {
+            state @ (ChainState::ConsumingSnapshot | ChainState::ConsumingUpstream(_, _)) => {
                 if matches!(new_state, ChainState::Done) {
                     self.done_count += 1;
                 }
                 *state = new_state;
             }
             ChainState::Done => panic!("should not report done multiple times"),
         }
+        self.calculate_progress();
```

Comment:
What about only calculating in-place when `gen_ddl_progress`, so that we don't need to maintain one more field?

Reply:
Actually I use it as a state to ensure the progress value will never go down, because the calculation itself doesn't guarantee the monotonicity.